tests/5.1/metadirective/test_metadirective_nothing.c

```diff
@@ -0,0 +1,116 @@
+//===--------------------- test_metadirective_nothing.c ---------------------===//
+//
+// OpenMP API Version 5.1 Nov 2020
+// 
+// Test for nothing directive within metadirectives. Runs a variety of
+// metadirectives that check if the nothing directive is properly rendered.
+// Primarily tests based on the fact that no matter what 'when' clause is 
+// rendered it should result in nothing, and thus no additional pragma should
+// be created. Thus, the threads should remain unchanged through this process
+// and the compiler should handle it properly. Handles both device and host
+// based tests.
+//
+////===---------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+
+#define N 1024
+
+int metadirectiveOnDevice() {
+   int errors = 0;
+   int scalar = 0;
+   int A[N];
+
+   for (int i = 0; i < N; i++) {
+      A[i] = 0;
+   }
+
+   #pragma omp metadirective \
+      when( device={kind(nohost)}: nothing ) \
+      when( device={arch("nvptx")}: nothing) \
+      when( device={arch("amd")}: nothing ) \
+      default( target map(to:from scalar, A) )
+      {
+         scalar = 10;
+         for (int i = 0; i < N; i++) {
+            A[i] = i + 2;
+         }
+      }
+
+   #pragma omp metadirective \
+      when( implementation={vendor(amd)}: nothing) \
+      when( implementation={vendor(nvidia)}: nothing) \
+      when( device={kind(nohost)}: nothing) \
+      default( target map(to:from scalar, A) )
+      {
+         scalar = 10;
+         for (int i = 0; i < N; i++) {
+            A[i] = i + 2;
+         }
+      }
+
+   OMPVV_INFOMSG("Test ran with a number of available devices greater than 0");
+   OMPVV_INFOMSG_IF(A[0] == 0, "Test recognized device was of arch/vendor/kind nvidia, amd, or nohost");
+   OMPVV_WARNING_IF(A[0] == 2 || A[0] == 4, "Test could not recognize if device was of arch/vendor/kind nvidia, amd or, nohost, even though there are devices available.");
```

Comment:
The first message I don't understand — it is also output for `omp_get_num_devices() == 0`.

Additionally: I note that `scalar` is not checked for – resulting in the warning (here and in the other function):
`warning: variable ‘scalar’ set but not used [-Wunused-but-set-variable]`

Reply:
Scalar is mostly redundant (since if the array is not changed we know the test worked) so I cleaned that up. 

However that first message is only run from the on device method. It should only appear whenever `omp_get_num_devices() > 0` per the if statement (line 101) in the `main` function. The first message is meant to let the user know that this test ran on device and not host, maybe it could be reworded to be more clear?

Reply:
> However that first message is only run from the on device method. It should only appear whenever `omp_get_num_devices() > 0` per the if statement (line 101) in the `main` function. The first message is meant to let the user know that this test ran on device and not host, maybe it could be reworded to be more clear?

Missed this – as I focused more on single lines, loosing the big context.

Regarding GCC 12: I think I have to disappoint you. While GCC 12 supports the **nothing** directive, it does not support metadirectives (missed the feature-freeze deadline); there is an initial implementation for metadirectives in the OG11 branch (on the GCC git; actual git name: 'devel/omp/gcc-11' + based on GCC 11, soon to be replaced by OG12 ). Thus, this has to wait for GCC 13/mainline.

However, this does not prevent me from doing some testing. :-)

* * *

Additionally, I think _either_ you _or_ I misunderstand how the nothing directive expands. In my understanding,
```C
   #pragma omp metadirective \
      when( device={kind(nohost)}: nothing ) \
      when( device={arch("nvptx")}: nothing) \
      when( device={arch("amd")}: nothing ) \
      default( target map(tofrom: A) )
      {
         for (int i = 0; i < N; i++)  A[i] = i + 2;
      }
```
becomes (when `nothing` matches) equivalent to
```C
#pragma omp nothing
      {
         for (int i = 0; i < N; i++)  A[i] = i + 2;
      }
```
which in turn is equivalent to
```C
      {
         for (int i = 0; i < N; i++)  A[i] = i + 2;
      }
```
Thus, the loop is still executed with `nothing` — but on the host and not on the device. Thus, with `tofrom`, I actually expect the same result for the code. 

However, if the result for `A` is the same,
` OMPVV_INFOMSG_IF(A[0] == 0, "Test recognized device was of arch/vendor/kind nvidia, amd, or nohost");`

will never trigger – either `omp metadirective` is not recognized (as with GCC <= 12 (+ current mainline/GCC13)) or if it is recognized, any of `nothing`, execution on the host or device execution (due to `tofrom` and/or shared memory) should give the loop value (`A[i] = i + 2`).

**Solution:** Maybe: call `omp_is_initial_device()` in the `{ ... }` block for checking? And/or use `firstprivate` instead of `tofrom`?

Reply:
Ah gotcha, well I appreciate you running the test on GCC :) Makes catching bugs a lot easier when you can actually (mostly) compile the code!

That said, I see what you're getting at. I think I got confused from an example in the 5.1 example doc (page 460) which in part looks like:
```C
#pragma omp metadirective when(user={condition(run_parallel)}: parallel)
{
  if(omp_in_parallel() == 1 && omp_get_thread_num() == 0)
    {printf("PASSED 2 of 3\n");}

  #pragma omp metadirective \
     when( construct={parallel}, \
     user={condition(unbalanced)}: for schedule(guided) private(b)) \
     when( construct={parallel} : for schedule(static))
  for (int i=0; i<n; i++) {a[i]=i; if(i==n-1) b=1;}
}
``` 
Which I interpreted as the block of code being run _if the directive resolved to a construct_ however thinking back this does not line up with how the nothing directive really works as you point out, so yes I think I assumed a bit too quickly how the directive really worked. Thank you for the awesome explanation!

`firstprivate` may be a way to go with the target directive, but maybe also using something like `omp_in_parallel` might work too to check and see if a parallel region was generated, or nothing (which would mean it'd be 0). Could be a cleaner approach possibly.